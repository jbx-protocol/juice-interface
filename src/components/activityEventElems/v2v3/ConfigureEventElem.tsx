import { t, Trans } from '@lingui/macro'
import { MinimalCollapse } from 'components/MinimalCollapse'
import RichNote from 'components/RichNote'
import { BigNumber } from 'ethers'
import {
  V2V3FundingCycle,
  V2V3FundingCycleMetadata,
} from 'models/v2v3/fundingCycle'

import FundingCycleDetails from 'components/v2v3/V2V3Project/V2V3FundingCycleSection/FundingCycleDetails'
import { PV_V2 } from 'constants/pv'
import { V2V3ProjectContext } from 'contexts/v2v3/Project/V2V3ProjectContext'
import { ProjectEventsQuery } from 'generated/graphql'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useContext } from 'react'
import { ActivityEvent } from '../ActivityElement'

export default function ConfigureEventElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['configureEvent']
  withProjectLink?: boolean
}) {
  if (!event) return null

  const { primaryETHTerminal } = useContext(V2V3ProjectContext)

  const { data: distributionLimit } = useProjectDistributionLimit({
    projectId: event.projectId,
    terminal: primaryETHTerminal,
    configuration: event.configuration?.toString(),
  })

  const fundingCycle: Partial<V2V3FundingCycle> = {
    duration: BigNumber.from(event.duration),
    weight: BigNumber.from(event.weight),
    discountRate: BigNumber.from(event.discountRate),
    ballot: event.ballot,
    start: event.mustStartAtOrAfter
      ? BigNumber.from(event.mustStartAtOrAfter)
      : undefined,
  }

  const fundingCycleMetadata: Partial<V2V3FundingCycleMetadata> = {
    reservedRate: BigNumber.from(event.reservedRate),
    redemptionRate: BigNumber.from(event.redemptionRate),
    global: {
      pauseTransfers: event.transfersPaused,
      allowSetController: event.setControllerAllowed,
      allowSetTerminals: event.setTerminalsAllowed,
    },
    allowControllerMigration: event.controllerMigrationAllowed,
    allowTerminalMigration: event.terminalMigrationAllowed,
    pausePay: event.pausePay,
    pauseRedeem: event.redeemPaused,
    pauseBurn: event.burnPaused,
    pauseDistributions: event.distributionsPaused,
    useTotalOverflowForRedemptions: event.useTotalOverflowForRedemptions,
    holdFees: event.shouldHoldFees,
    dataSource: event.dataSource,
    useDataSourceForPay: event.useDataSourceForPay,
    useDataSourceForRedeem: event.useDataSourceForRedeem,
    allowMinting: event.mintingAllowed,
  }

  return (
    <ActivityEvent
      event={event}
      header={t`Edited cycle`}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      subject={
        <MinimalCollapse
          header={
            <span className="font-normal">
              <Trans>Details</Trans>
            </span>
          }
        >
          <FundingCycleDetails
            fundingCycle={fundingCycle as V2V3FundingCycle}
            fundingCycleMetadata={
              fundingCycleMetadata as V2V3FundingCycleMetadata
            }
            distributionLimit={distributionLimit?.[0]}
            distributionLimitCurrency={distributionLimit?.[1]}
            mintRateZeroAsUnchanged
          />
          {event.memo ? <RichNote className="mt-3" note={event.memo} /> : null}
        </MinimalCollapse>
      }
    />
  )
}
