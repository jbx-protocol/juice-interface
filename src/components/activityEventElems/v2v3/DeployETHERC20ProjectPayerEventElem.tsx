import { t, Trans } from '@lingui/macro'
import EthereumAddress from 'components/EthereumAddress'
import RichNote from 'components/RichNote'
import { ProjectEventsQuery } from 'generated/graphql'

import { PV_V2 } from 'constants/pv'
import { ActivityEvent } from '../ActivityElement'

export default function DeployETHERC20ProjectPayerEventElem({
  event,
  withProjectLink,
}: {
  event:
    | ProjectEventsQuery['projectEvents'][0]['deployETHERC20ProjectPayerEvent']
  withProjectLink?: boolean
}) {
  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      header={t`Deployed a project payer address`}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      subject={
        <Trans>
          from <EthereumAddress address={event.from} />
        </Trans>
      }
      extra={
        <div>
          <Trans>
            Address: <EthereumAddress address={event.address} />
          </Trans>
          {event.memo && (
            <div className="mt-2">
              <RichNote note={event.memo} />
            </div>
          )}
        </div>
      }
    />
  )
}
