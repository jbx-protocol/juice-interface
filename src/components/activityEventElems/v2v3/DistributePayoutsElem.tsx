import { t } from '@lingui/macro'
import EthereumAddress from 'components/EthereumAddress'
import { JuiceboxAccountLink } from 'components/JuiceboxAccountLink'
import ETHAmount from 'components/currency/ETHAmount'
import V2V3ProjectHandleLink from 'components/v2v3/shared/V2V3ProjectHandleLink'
import { PV_V2 } from 'constants/pv'
import {
  ProjectEventsQuery,
  useSplitDistributionsForDistributePayoutsEventQuery,
} from 'generated/graphql'
import { client } from 'lib/apollo/client'

import { ActivityEvent } from '../ActivityElement'

export default function DistributePayoutsElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['distributePayoutsEvent']
  withProjectLink?: boolean
}) {
  // Load individual DistributeToPayoutSplit events, emitted by internal transactions of the DistributeReservedPayouts transaction
  const { data } = useSplitDistributionsForDistributePayoutsEventQuery({
    client,
    variables: {
      distributePayoutsEvent: event?.id,
    },
  })

  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      header={t`Sent payouts`}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      subject={
        data?.distributeToPayoutSplitEvents.length ? (
          <span className="font-heading text-lg">
            <ETHAmount amount={event.distributedAmount} />
          </span>
        ) : null
      }
      extra={
        <div>
          {data?.distributeToPayoutSplitEvents.map(e => (
            <div
              key={e.id}
              className="flex items-baseline justify-between text-sm"
            >
              <div className="font-medium">
                {e.splitProjectId ? (
                  <V2V3ProjectHandleLink
                    className="text-grey-900 dark:text-slate-100"
                    projectId={e.splitProjectId}
                  />
                ) : (
                  <JuiceboxAccountLink
                    className="text-grey-900 dark:text-slate-100"
                    address={e.beneficiary}
                  />
                )}
                :
              </div>

              <div className="text-secondary">
                <ETHAmount amount={e.amount} />
              </div>
            </div>
          ))}

          {event.beneficiaryDistributionAmount?.gt(0) && (
            <div className="flex items-baseline justify-between">
              <div>
                <EthereumAddress
                  className="text-grey-900 dark:text-slate-100"
                  address={event.beneficiary}
                />
                :
              </div>
              <div className="text-sm text-grey-500 dark:text-grey-300">
                <ETHAmount amount={event.beneficiaryDistributionAmount} />
              </div>
            </div>
          )}
        </div>
      }
    />
  )
}
