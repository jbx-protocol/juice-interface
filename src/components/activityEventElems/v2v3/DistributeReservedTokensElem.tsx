import { Trans } from '@lingui/macro'
import EthereumAddress from 'components/EthereumAddress'
import { JuiceboxAccountLink } from 'components/JuiceboxAccountLink'
import { PV_V2 } from 'constants/pv'
import { V1ProjectContext } from 'contexts/v1/Project/V1ProjectContext'
import {
  ProjectEventsQuery,
  useSplitDistributionsForDistributeReservedTokensEventQuery,
} from 'generated/graphql'
import { client } from 'lib/apollo/client'
import { useContext } from 'react'
import { formatWad, fromWad } from 'utils/format/formatNumber'
import { tokenSymbolText } from 'utils/tokenSymbolText'
import { ActivityEvent } from '../ActivityElement'

export default function DistributeReservedTokensEventElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['distributeReservedTokensEvent']
  withProjectLink?: boolean
}) {
  const { tokenSymbol } = useContext(V1ProjectContext)

  // Load individual DistributeToReservedTokenSplit events, emitted by internal transactions of the DistributeReservedTokens transaction
  const { data } = useSplitDistributionsForDistributeReservedTokensEventQuery({
    client,
    variables: {
      distributeReservedTokensEvent: event?.id,
    },
  })

  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      header={
        <Trans>
          Sent reserved{' '}
          {tokenSymbolText({
            tokenSymbol,
            capitalize: false,
            plural: true,
          })}
        </Trans>
      }
      subject={
        <div className="font-heading text-lg">
          {formatWad(event.tokenCount, { precision: 0 })}{' '}
          {tokenSymbolText({
            tokenSymbol,
            capitalize: false,
            plural: parseInt(fromWad(event.tokenCount) || '0') !== 1,
          })}
        </div>
      }
      extra={
        <div>
          {data?.distributeToReservedTokenSplitEvents.map(e => (
            <div key={e.id} className="flex items-baseline justify-between">
              <div>
                <JuiceboxAccountLink
                  className="text-grey-900 dark:text-slate-100"
                  address={e.beneficiary}
                />
              </div>

              <div className="text-sm text-grey-500 dark:text-grey-300">
                {formatWad(e.tokenCount, { precision: 0 })}{' '}
                {tokenSymbolText({
                  tokenSymbol,
                  capitalize: false,
                  plural: true,
                })}
              </div>
            </div>
          ))}

          {event.beneficiaryTokenCount?.gt(0) && (
            <div className="flex items-baseline justify-between">
              <div>
                <EthereumAddress
                  className="text-grey-900 dark:text-slate-100"
                  address={event.beneficiary}
                />
                :
              </div>
              <div className="text-sm text-grey-500 dark:text-grey-300">
                {formatWad(event.beneficiaryTokenCount, {
                  precision: 0,
                })}
              </div>
            </div>
          )}
        </div>
      }
    />
  )
}
