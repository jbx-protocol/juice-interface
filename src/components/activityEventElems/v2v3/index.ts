export * from './ConfigureEventElem'
export * from './DeployETHERC20ProjectPayerEventElem'
export * from './DistributePayoutsElem'
export * from './DistributeReservedTokensElem'
export * from './SetFundAccessConstraintsEventElem'
