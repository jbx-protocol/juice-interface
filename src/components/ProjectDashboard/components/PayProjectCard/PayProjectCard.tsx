import { QuestionMarkCircleIcon } from '@heroicons/react/24/outline'
import { Trans, t } from '@lingui/macro'
import { Button, Tooltip } from 'antd'
import { usePayProjectCard } from 'components/ProjectDashboard/hooks'
import { Formik } from 'formik'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { twMerge } from 'tailwind-merge'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { DisplayCard } from '../ui'
import { PayInput } from './components/PayInput'
import { TokensPerEth } from './components/TokensPerEth'

export const PayProjectCard = ({ className }: { className?: string }) => {
  const { validationSchema, paymentsPaused, addPay } = usePayProjectCard()
  return (
    <DisplayCard
      className={twMerge('flex flex-col gap-2 px-4 md:pr-9 md:pl-6', className)}
    >
      <div className="text-base font-medium">
        <Tooltip
          className="inline-flex items-center gap-1"
          title={<Trans>Send ETH payments to this project.</Trans>}
        >
          <span>
            <Trans>Pay Project</Trans>
          </span>
          <QuestionMarkCircleIcon className="h-4 w-4 text-grey-500 dark:text-slate-200" />
        </Tooltip>
      </div>
      <Formik
        initialValues={{
          payAmount: {
            amount: undefined as unknown as number,
            currency: V2V3_CURRENCY_ETH as V2V3CurrencyOption,
          },
        }}
        validationSchema={validationSchema}
        onSubmit={addPay}
      >
        {props => (
          <form className="flex flex-col gap-2" onSubmit={props.handleSubmit}>
            <div className="flex flex-col gap-2 md:flex-row">
              <PayInput
                className="h-12 md:flex-1"
                placeholder="0"
                value={{
                  amount: props.values.payAmount.amount?.toString() ?? '',
                  currency: props.values.payAmount.currency,
                }}
                onChange={v => props.setFieldValue('payAmount', v)}
                onBlur={props.handleBlur}
                name="payAmount"
              />
              <Tooltip
                title={t`Payments to this project are paused in this cycle.`}
                open={paymentsPaused ? undefined : false}
                className="h-12"
              >
                <Button
                  loading={paymentsPaused === undefined}
                  disabled={paymentsPaused}
                  htmlType="submit"
                  className="h-12 text-base"
                  style={{ height: '48px' }}
                  type="primary"
                >
                  <Trans>Add payment</Trans>
                </Button>
              </Tooltip>
            </div>
            <div
              data-testid="pay-project-card-tokens-per-pay"
              className="text-xs text-smoke-500 dark:text-slate-200"
            >
              <TokensPerEth currencyAmount={props.values.payAmount} />
            </div>
            {props.errors.payAmount?.amount && props.touched.payAmount && (
              <div className="text-error-500">
                {props.errors.payAmount.amount}
              </div>
            )}
          </form>
        )}
      </Formik>
    </DisplayCard>
  )
}
