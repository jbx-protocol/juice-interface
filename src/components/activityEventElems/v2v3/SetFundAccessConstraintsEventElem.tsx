import { t } from '@lingui/macro'
import { AmountInCurrency } from 'components/currency/AmountInCurrency'
import { ProjectEventsQuery } from 'generated/graphql'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { V2V3CurrencyName } from 'utils/v2v3/currency'
import { MAX_DISTRIBUTION_LIMIT } from 'utils/v2v3/math'

import { PV_V2 } from 'constants/pv'
import { ActivityEvent } from '../ActivityElement'

export default function SetFundAccessConstraintsEventElem({
  event,
  withProjectLink,
}: {
  event: ProjectEventsQuery['projectEvents'][0]['setFundAccessConstraintsEvent']
  withProjectLink?: boolean
}) {
  if (!event) return null

  return (
    <ActivityEvent
      event={event}
      withProjectLink={withProjectLink}
      pv={PV_V2}
      header={t`Edited payout`}
      subject={
        <div className="font-heading text-lg">
          {event.distributionLimit.eq(MAX_DISTRIBUTION_LIMIT) ? (
            t`Unlimited`
          ) : (
            <AmountInCurrency
              amount={event.distributionLimit}
              currency={V2V3CurrencyName(
                event.distributionLimitCurrency as V2V3CurrencyOption,
              )}
            />
          )}
        </div>
      }
    />
  )
}
