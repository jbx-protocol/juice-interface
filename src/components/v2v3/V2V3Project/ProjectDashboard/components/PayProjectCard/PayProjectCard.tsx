import { useMemo } from 'react'

import {
  NoSymbolIcon,
  QuestionMarkCircleIcon,
} from '@heroicons/react/24/outline'
import { Trans, t } from '@lingui/macro'
import { Button, Tooltip } from 'antd'
import { Formik } from 'formik'
import { useV2BlockedProject } from 'hooks/useBlockedProject'
import { V2V3CurrencyOption } from 'models/v2v3/currencyOption'
import { twMerge } from 'tailwind-merge'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import {
  usePayProjectCard,
  useProjectMetadata,
  useProjectOFACContext,
} from 'components/v2v3/V2V3Project/ProjectDashboard/hooks'
import { DisplayCard } from '../ui/DisplayCard'
import { PayInput } from './components/PayInput'
import { TokensPerEth } from './components/TokensPerEth'

export const PayProjectCard = ({ className }: { className?: string }) => {
  const isBlockedProject = useV2BlockedProject()
  const { projectMetadata } = useProjectMetadata()
  const { isAddressListedInOFAC } = useProjectOFACContext()
  const { validationSchema, paymentsPaused, addPay } = usePayProjectCard()
  const determiningIfProjectCanReceivePayments = paymentsPaused === undefined

  const isJuicecrowdProject = useMemo(
    () => projectMetadata?.domain === 'juicecrowd',
    [projectMetadata?.domain],
  )

  const isPaymentDisabled = useMemo(() => {
    return (
      paymentsPaused ||
      isBlockedProject ||
      isAddressListedInOFAC ||
      isJuicecrowdProject
    )
  }, [
    isAddressListedInOFAC,
    paymentsPaused,
    isBlockedProject,
    isJuicecrowdProject,
  ])

  const paymentTooltip = useMemo(() => {
    if (paymentsPaused || isJuicecrowdProject) {
      return {
        title: t`Payments to this project are paused in this cycle.`,
        open: paymentsPaused || isJuicecrowdProject ? undefined : false,
      }
    }

    if (isAddressListedInOFAC) {
      return {
        title: t`Unfortunately, this action cannot be completed because your wallet address failed compliance check`,
        open: true,
      }
    }

    return {
      open: false,
    }
  }, [paymentsPaused, isAddressListedInOFAC, isJuicecrowdProject])

  return (
    <DisplayCard
      className={twMerge('flex flex-col gap-2 px-4 md:pl-6 md:pr-9', className)}
    >
      <div className="text-base font-medium">
        <Tooltip
          className="inline-flex items-center gap-1"
          title={<Trans>Send ETH payments to this project.</Trans>}
        >
          <span>
            <Trans>Pay Project</Trans>
          </span>
          <QuestionMarkCircleIcon className="h-4 w-4 text-grey-500 dark:text-slate-200" />
        </Tooltip>
      </div>
      {paymentsPaused || isJuicecrowdProject ? (
        <div className="text-grey-600 dark:text-slate-200">
          <div className="flex cursor-not-allowed items-center gap-2 rounded-lg bg-smoke-100 px-4 py-3 text-base leading-none dark:bg-slate-800">
            <NoSymbolIcon className="h-5 w-5" />
            <Trans>This project has disabled payments</Trans>
          </div>
          <div className="mt-3 text-sm">
            <Trans>Tokens cannot be issued while payments are disabled.</Trans>
          </div>
        </div>
      ) : (
        <Formik
          initialValues={{
            payAmount: {
              amount: undefined as unknown as number,
              currency: V2V3_CURRENCY_ETH as V2V3CurrencyOption,
            },
          }}
          validationSchema={validationSchema}
          onSubmit={addPay}
        >
          {props => (
            <form className="flex flex-col gap-2" onSubmit={props.handleSubmit}>
              <div className="flex flex-col gap-2 md:flex-row">
                <PayInput
                  className="h-12 md:flex-1"
                  placeholder="0"
                  value={{
                    amount: props.values.payAmount.amount?.toString() ?? '',
                    currency: props.values.payAmount.currency,
                  }}
                  onChange={v => props.setFieldValue('payAmount', v)}
                  onBlur={props.handleBlur}
                  name="payAmount"
                />
                <Tooltip className="h-12" {...paymentTooltip}>
                  <Button
                    loading={determiningIfProjectCanReceivePayments}
                    disabled={isPaymentDisabled}
                    htmlType="submit"
                    className="h-12 text-base"
                    style={{ height: '48px' }}
                    type="primary"
                  >
                    <Trans>Add payment</Trans>
                  </Button>
                </Tooltip>
              </div>
              <div
                data-testid="pay-project-card-tokens-per-pay"
                className="text-xs text-smoke-500 dark:text-slate-200"
              >
                <TokensPerEth currencyAmount={props.values.payAmount} />
              </div>
              {props.errors.payAmount?.amount && props.touched.payAmount && (
                <div className="text-error-500">
                  {props.errors.payAmount.amount}
                </div>
              )}
            </form>
          )}
        </Formik>
      )}
    </DisplayCard>
  )
}
