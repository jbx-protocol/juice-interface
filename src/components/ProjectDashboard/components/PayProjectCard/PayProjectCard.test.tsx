/**
 * @jest-environment jsdom
 */
/* eslint-disable @typescript-eslint/no-unused-vars */
import { fireEvent, render, waitFor } from '@testing-library/react'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { PayProjectCard } from './PayProjectCard'

jest.mock('./components/TokensPerEth', () => ({
  TokensPerEth: jest
    .fn()
    .mockImplementation(currencyAmount => (
      <div data-testid="tokens-per-eth">{JSON.stringify(currencyAmount)}</div>
    )),
}))

jest.mock('./components/PayInput')

describe('PayProjectCard', () => {
  it('renders', async () => {
    const { container } = render(<PayProjectCard />)
    await waitFor(() => expect(container).toMatchSnapshot(), { timeout: 10000 })
  })

  test('entering text into input updates token value', async () => {
    const { getByTestId, getByRole } = render(<PayProjectCard />)
    const input = getByRole('textbox')
    const tokensPerEth = getByTestId('tokens-per-eth')
    await waitFor(
      () => {
        expect(input).toHaveValue('')
        expect(tokensPerEth).toHaveTextContent(
          JSON.stringify({
            currencyAmount: { amount: undefined, currency: V2V3_CURRENCY_ETH },
          }),
        )
      },
      { timeout: 10000 },
    )
    await waitFor(
      () => {
        fireEvent.change(input, { target: { value: '1' } })
        expect(input).toHaveValue('1')
        expect(tokensPerEth).toHaveTextContent(
          JSON.stringify({
            currencyAmount: { amount: 1, currency: V2V3_CURRENCY_ETH },
          }),
        )
      },
      { timeout: 10000 },
    )
  })
})
