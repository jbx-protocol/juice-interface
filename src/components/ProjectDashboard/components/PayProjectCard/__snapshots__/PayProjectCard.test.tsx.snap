// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayProjectCard renders 1`] = `
<div>
  <div
    class="rounded-lg bg-smoke-50 py-5 dark:bg-slate-700 flex flex-col gap-2 px-4 md:pr-9 md:pl-6"
  >
    <div
      class="text-base font-medium"
    >
      <span
        class="inline-flex items-center gap-1"
      >
        <span>
          Pay Project
        </span>
        <svg
          aria-hidden="true"
          class="h-4 w-4 text-grey-500 dark:text-slate-200"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.879 7.519c1.171-1.025 3.071-1.025 4.242 0 1.172 1.025 1.172 2.687 0 3.712-.203.179-.43.326-.67.442-.745.361-1.45.999-1.45 1.827v.75M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9 5.25h.008v.008H12v-.008z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
    </div>
    <form
      class="flex flex-col gap-2"
    >
      <div
        class="flex flex-col gap-2 md:flex-row"
      >
        <input
          class="h-12 md:flex-1"
          data-testid="pay-input"
          name="payAmount"
          placeholder="0"
          value=""
        />
        <button
          class="ant-btn ant-btn-primary ant-btn-loading h-12 text-base h-12"
          style="height: 48px;"
          type="submit"
        >
          <span
            class="ant-btn-loading-icon"
          >
            <span
              aria-label="loading"
              class="anticon anticon-loading anticon-spin"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="loading"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                />
              </svg>
            </span>
          </span>
          Add payment
        </button>
      </div>
      <div
        class="text-xs text-smoke-500 dark:text-slate-200"
        data-testid="pay-project-card-tokens-per-pay"
      >
        <div
          data-testid="tokens-per-eth"
        >
          {"currencyAmount":{"currency":1}}
        </div>
      </div>
    </form>
  </div>
</div>
`;
